private/bufpkg/bufimage/source_retention_options.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufimage
+
+import (
+	"fmt"
+
+	imagev1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/image/v1"
+	"github.com/bufbuild/buf/private/pkg/protodescriptor"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// StripSourceRetentionOptionsFromFile returns a file descriptor proto that omits any
+// options that in file that are defined to be retained only in source. If file has no
+// such options, then it is returned as is. If it does have such options, a copy is
+// made; the given file will not be mutated.
+//
+// The given file must be either a *descriptorpb.FileDescriptorProto or a
+func StripSourceRetentionOptionsFromFile[F protodescriptor.FileDescriptor](
+	file F,
+) (F, error) {
+	var zero F
+	var dirty bool
+	newOpts, err := stripSourceRetentionOptions(file.GetOptions())
+	if err != nil {
+		return zero, err
+	}
+	if newOpts != file.GetOptions() {
+		dirty = true
+	}
+	newMsgs, changed, err := updateAll(file.GetMessageType(), stripSourceRetentionOptionsFromMessage)
+	if err != nil {
+		return zero, err
+	}
+	if changed {
+		dirty = true
+	}
+	newEnums, changed, err := updateAll(file.GetEnumType(), stripSourceRetentionOptionsFromEnum)
+	if err != nil {
+		return zero, err
+	}
+	if changed {
+		dirty = true
+	}
+	newExts, changed, err := updateAll(file.GetExtension(), stripSourceRetentionOptionsFromField)
+	if err != nil {
+		return zero, err
+	}
+	if changed {
+		dirty = true
+	}
+	newSvcs, changed, err := updateAll(file.GetService(), stripSourceRetentionOptionsFromService)
+	if err != nil {
+		return zero, err
+	}
+	if changed {
+		dirty = true
+	}
+
+	if !dirty {
+		return file, nil
+	}
+
+	newFile, err := shallowCopy(file)
```

Comment:
What is shallow about the copy exactly? The method documentation implies that I'm getting a full copy back, which would say to me I can do whatever I want with the return value of the method, but this would imply otherwise I think

Reply:
I can update the doc comments. It definitely wasn't meant to suggest that you can do whatever you want the value but rather that it does not mutate the incoming value.

It copies only the elements that changed, which are leaf messages that were mutated and their ancestors in the hierarchy. (So if any leaf is changed, you'll get back a different `*descriptorpb.FileDescriptorProto`, but it can still share structure with the input.)